storage: add Replica state assertions

tbg opened the pull request:
In particular, we check that an initialized Replica always has a complete
state, and that no empty HardState is persisted.

I don't think we're going to find much real-world use for these assertions,
but they did help clean up some tests and establish clarity about when what
state is persisted in reality.

Removed `raftTruncatedState` (which had been superseded by `loadTruncatedState`
but was still used).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7379)

<!-- Reviewable:end -->

tbg committed:
storage: add existence return value to load(.*)

In preparation for asserting on existence in various locations.

tschottdorf commented:
Thanks Kenji. I'll take a look tomorrow, but just wanted to point out that this change may require a stop-the-world migration.

tschottdorf commented:
Ha, the [test failures](https://circleci.com/gh/tschottdorf/cockroach/2353) highlight the obvious omission: Old versions were way less stringent about writing their state, so the reference tests fail on the assertions.

The previous state refactors (writing the initial states) would also have required a stop-the-world upgrade (the divergence could probably last, since new versions still don't write every key at least once, for example the frozen status).

tamird commented:
Reviewed 11 of 11 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica_state.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KKrHbe0-qm6AL4nuqdf:-KKrHbe0-qm6AL4nuqdg:1413706000) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae332635aef4d179a46fdd48f400bce70daea1a/storage/replica_state.go#L48)):*
> ```Go
> 	assert := func(curExists bool) {
> 		if exists != curExists {
> 			panic(util.ErrorfSkipFrames(1,
> ```

why panic rather than Fatal?

---

*[storage/replica_state.go, line 64 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KKrHmK_JOmN6duNufxS:-KKrHmK_JOmN6duNufxT:771743198) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae332635aef4d179a46fdd48f400bce70daea1a/storage/replica_state.go#L64)):*
> ```Go
> 	assert(curExists)
> 
> 	if _, s.GCThreshold, err = loadGCThreshold(reader, desc.RangeID); err != nil {
> ```

did you mean to assign curExists at every one of these lines?

---

*[storage/engine/mvcc.go, line 416 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KKrGvAmOZdOckjStrLU:-KKrGvAmOZdOckjStrLV:-299045963) ([raw file](https://github.com/cockroachdb/cockroach/blob/07dede2ff71ca2adfa65b1f7c1c58da8469901a1/storage/engine/mvcc.go#L416)):*
> ```Go
> // sets the values in the enginepb.MVCCStats struct.
> //
> // TODO(tschottdorf): see whether keeping this method around makes sense.
> ```

can you elaborate on why it might not? this looks like something easy to forget

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 14 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica_state.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KKrHbe0-qm6AL4nuqdf:-KLGMy-Vn6oXgkAYulwZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae332635aef4d179a46fdd48f400bce70daea1a/storage/replica_state.go#L48)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why panic rather than Fatal?
</details>
Done.

---

*[storage/replica_state.go, line 64 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KKrHmK_JOmN6duNufxS:-KLGMzBYszc4t0SQDGEs:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ae332635aef4d179a46fdd48f400bce70daea1a/storage/replica_state.go#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did you mean to assign curExists at every one of these lines?
</details>
Done.

---

*[storage/engine/mvcc.go, line 416 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KKrGvAmOZdOckjStrLU:-KLGN5paX717kjOMjmQw:2072458273) ([raw file](https://github.com/cockroachdb/cockroach/blob/07dede2ff71ca2adfa65b1f7c1c58da8469901a1/storage/engine/mvcc.go#L416)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you elaborate on why it might not? this looks like something easy to forget
</details>
Removed. I had added the comment since it seemed silly to have a special-cased method like that (we don't for anything else), but stats computation is a little special since we pushed it down to C++.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 10 of 10 files at r3, 9 of 9 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[acceptance/reference_test.go, line 49 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KLH9_spm3HLD_oq1Ofj:-KLH9_spm3HLD_oq1Ofk:1214612439) ([raw file](https://github.com/cockroachdb/cockroach/blob/cebdc8faba8a4c6917965986757e07b2f7bc795f/acceptance/reference_test.go#L49)):*
> ```Go
> function finish() {
>   touch oldout newout
>   cat oldout
> ```

nit `cat oldout newout`

---

*[acceptance/reference_test.go, line 60 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KLH9eBVh5AqIqMVtnnT:-KLH9eBVh5AqIqMVtnnU:-499140523) ([raw file](https://github.com/cockroachdb/cockroach/blob/cebdc8faba8a4c6917965986757e07b2f7bc795f/acceptance/reference_test.go#L60)):*
> ```Go
> bin=/%s/cockroach
> # TODO(bdarnell): when --background is in referenceBinPath, use it here and below.
> $bin start --alsologtostderr & &> oldout
> ```

is this the right syntax? what does `& &>` do?

---

*[storage/store.go, line 814 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KLHAKOEUdg1ILhWXLW1:-KLHAKOEUdg1ILhWXLW2:1066326232) ([raw file](https://github.com/cockroachdb/cockroach/blob/cebdc8faba8a4c6917965986757e07b2f7bc795f/storage/store.go#L814)):*
> ```Go
> 	}
> 	batch := s.engine.NewBatch()
> 	state, err := loadState(batch, &desc, false /* !assert */)
> ```

short comment here plz

---

*[storage/engine/mvcc.go, line 416 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KKrGvAmOZdOckjStrLU:-KLHAW33YpH034Zsh9m3:804825280) ([raw file](https://github.com/cockroachdb/cockroach/blob/07dede2ff71ca2adfa65b1f7c1c58da8469901a1/storage/engine/mvcc.go#L416)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Removed. I had added the comment since it seemed silly to have a special-cased method like that (we don't for anything else), but stats computation is a little special since we pushed it down to C++.
</details>
I don't completely follow, but OK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: add Replica state assertions

In particular, we check that an initialized Replica always has a complete
state, and that no empty HardState is persisted.

tschottdorf commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[acceptance/reference_test.go, line 49 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KLH9_spm3HLD_oq1Ofj:-KLHkf2Kqq0dwGeQINX5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cebdc8faba8a4c6917965986757e07b2f7bc795f/acceptance/reference_test.go#L49)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit `cat oldout newout`
</details>
Done.

---

*[acceptance/reference_test.go, line 60 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KLH9eBVh5AqIqMVtnnT:-KLHkgzqR_-Y0fV9dSx0:1926338162) ([raw file](https://github.com/cockroachdb/cockroach/blob/cebdc8faba8a4c6917965986757e07b2f7bc795f/acceptance/reference_test.go#L60)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this the right syntax? what does `& &>` do?
</details>
`&` launches in the background, `&>` redirects everything.

---

*[storage/store.go, line 814 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-KLHAKOEUdg1ILhWXLW1:-KLHl-8HjFsWIeT8qS8s:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cebdc8faba8a4c6917965986757e07b2f7bc795f/storage/store.go#L814)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> short comment here plz
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-:-KLHlpRekpmQpEhoajKv:1417088721)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 3 of 14 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7379#-:-KLc12PH6oSohjG35qwH:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I extracted everything I thought was worth keeping: #7828 #7826 #7827. Closing.
